group_vars/all.yml.sample

```diff
@@ -690,8 +690,8 @@ dummy:
 # By default, keys will be auto-generated.
 #
 #openstack_keys:
-#  - { name: client.glance, caps: { mon: "profile rbd", osd: "profile rbd pool=volumes, profile rbd pool={{ openstack_glance_pool.name }}, osd blacklist"}, mode: "0600" }
-#  - { name: client.cinder, caps: { mon: "profile rbd", osd: "profile rbd pool={{ openstack_cinder_pool.name }}, profile rbd pool={{ openstack_nova_pool.name }}, profile rbd pool={{ openstack_glance_pool.name }}, osd blacklist"}, mode: "0600" }
+#  - { name: client.glance, caps: { mon: "profile rbd", osd: "allow class-read object_prefix rbd_children, profile rbd pool={{ openstack_glance_pool.name }}, osd blacklist"}, mode: "0600" }
+#  - { name: client.cinder, caps: { mon: "profile rbd", osd: "allow class-read object_prefix rbd_children, profile rbd pool={{ openstack_cinder_pool.name }}, profile rbd pool={{ openstack_nova_pool.name }}, profile rbd pool={{ openstack_glance_pool.name }}, osd blacklist"}, mode: "0600" }
```

Comment:
'class-read object_prefix rbd_children' is implicitly allowed by 'profile rbd' so it is not necessary

Reply:
Are you sure about that? Since I didn't find anything related to class-read for `profile rbd` in [docs](https://docs.ceph.com/docs/nautilus/rados/operations/user-management/):

> When used as an OSD cap, it provides read-write access to an RBD client application.

So according to this I'd say that it's regular `rwx` privileges. And without class-read I was having issues with creating volumes from images.

Reply:
pretty sure it includes class-read yes [1] [2]

1. https://github.com/ceph/ceph/blob/nautilus/src/osd/OSDCap.cc#L335
2. https://github.com/ceph/ceph/blob/nautilus/src/osd/OSDCap.h#L42

Reply:
Yep, seems that you're right... Ok, I've dropped specification of object class out of commit.

Reply:
@gfidente actually setting `allow class-read object_prefix rbd_children` is very necessary on all openstack keys.

Let me use `client.glance` as an example. When you create a glance image and then build CoW disks from it in nova, the image is protected in ceph. Once you decide to delete an image in glance, glance will first try to unprotect the image, which results in ceph scanning all pools for CoW children of the glance image. This process fails without `allow class-read object_prefix rbd_children` since glance does not have the ability to introspect all pools on the cluster to find child disks, and the unprotect fails, so the image cannot be deleted.

This is also true in nova (for nova live snapshots), and in cinder for volume snapshots.

All openstack keys should have `allow class-read object_prefix rbd_children` set.

Reply:
sure it is, what I am saying is that it's a subset of the permissions granted by the 'profile rbd' setting

Reply:
The 'profile rbd' is scoped to a specific pool

Reply:
The problem is the pools that the key does not have `profile rbd` access to

Reply:
I think I understand but I am not sure this is what happens; for example nova does not have permissions to read images in the glance pool either.

Instead, I think nova will download the glance image via API and populate the guests' disk with its cached copy. It does not need permissions to read into the glance backend and the image snapshots are created by glance and stay in the glance pool.

To be able to use cinder volumes instead, libvirt needs to be configured with the cinder secret so that nova can attach guests to images found in the cinder pool. Similarily to what happens for glance images, volume snapshots stay in the cinder pool.

Reply:
Framing this another way, forget about OpenStack for a minute. You've got OpenStack running and all is well.

Now you decide to create another pool `foobar` on your ceph cluster for non-OpenStack uses. Maybe iSCSI shares or something. OpenStack services don't know about or use this pool, but suddenly glance can no longer delete disks from its own `images` pool because when it attempts to unprotect a disk, it tries to scan `foobar` (along with all other pools on the cluster) for children, but fails due to the missing `class-read object_prefix rbd_children` permission.

Reply:
> I think I understand but I am not sure this is what happens; for example nova does not have permissions to read images in the glance pool either.
> 

Actually it does. That is how you can build rbd children from the images pool. See https://github.com/ceph/ceph-ansible/blob/70cf2a5846e38e98b32c47a176cb38488af3dc12/group_vars/all.yml.sample#L694 specifically `profile rbd pool={{ openstack_glance_pool.name }}`

The issue here is not image creates. The issue is unprotecting parent images on clusters where non-OpenStack pools exist (RGW, etc.).

When a service (glance, cinder, nova, whatever) tries to unprotect an image, *all* pools on the cluster are scanned for children of the image you are unprotecting. If some pools cannot be scanned (ie. because the `class-read object_prefix rbd_children` permission is scoped to specific pools only, rather than cluster-wide), then the unprotect operation fails and the rbd cannot be deleted.

Reply:
adding class-read on rbd-children for the glance keyring is fine and certainly the least disruptive change but I wanted to share my rationale on why it might make more sense to use a single keyring for cinder/glance/nova then

1) to be able to create clones into another pool, the other consumer (for example nova) must first have read permissions on the glance pool

2) nova not only reads from the glance pool to create COW clones but it also tries to write in the glance pool when shelving guests

3) nova (libvirt) also needs to know the cinder secret to be able to start guests from cinder volumes

4) cinder behaves like nova

ceph-ansible is currently not creating a keyring for nova, probably assuming nova will be configurd to use the cinder keyring for its ephemeral storage as well? if so it is already sharing its keyring with glance, what point having a separate keyring for glance only?

Reply:
@gfidente if you want to use the same key for all 3 services, it looks like `client.openstack` does what you're proposing.

@noonedeadpunk we probably want to add `class-read object_prefix rbd_children` to that `client.openstack` key also I'm guessing.

Reply:
@logan2211 I am used to deploy a single keyring indeed; maybe I see little benefit in having a shared keyring for nova/cinder with all the permissions to write in the glance pool too ... and then a different one for glance only with permissions to scan all pools for children objects. It seems to have the only benefit of preventing glance (and glance only) from using the cinder and nova pools.

Adding class-read for children to the client.openstack keyring surely makes sense in a scenario where you assume other tools have read permissions on the openstack pools and could, hence, create clones in their own pools. But this seems to me more a specific deployment decision, not something I would assume in a default setting.

Reply:
Noticed we ended up adding class-read on children for the cinder key as well, even though it does have enough permissions on the glance and nova pools because of the profile.

This goes a bit into the scenario where there could be other consumers with read permissions on the cinder pool creating clones elsewhere.

Reply:
Dropped class-read for cinder.